### PR TITLE
Change nested foreach ifs to early continue rector  bug when there is an or and all are true

### PR DESCRIPTION
**Pull request by derrickschoen**

### CODE REVIEW

File: rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_OR_both_truthy.php.inc [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    public function run()
+    {
+        $partPackageList = [ new \StdClass(), 'string' ];
+        $payload = [];
+
+        foreach( $partPackageList as $partPackage )
+        {
+            if(! is_null( $partPackage ) && ! is_object($partPackage) )
+            {
+                if (1 === 1 || 2 === 2)
```
Comment (TomasVotruba) [inline rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_OR_both_truthy.php.inc:18]: I think this `if` is enough to reproduce. The `if` above is not needed. Am I right?

Comment (derrickschoen) [reply rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_OR_both_truthy.php.inc:18]: ChangeNestedForeachIfsToEarlyContinueRector requires at least two levels of ifs to activate when I tried it in Rector v0.7.34

Comment (derrickschoen) [reply rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_OR_both_truthy.php.inc:18]: rules/solid/src/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector.php: L#99

```
    public function refactor(Node $node): ?Node
    {
        $nestedIfsWithOnlyNonReturn = $this->ifManipulator->collectNestedIfsWithNonBreaking($node);
        if (count($nestedIfsWithOnlyNonReturn) < 2) {
            return null;
        }

        return $this->processNestedIfsWithNonBreaking($node, $nestedIfsWithOnlyNonReturn);
    }
```

Comment (TomasVotruba) [reply rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_OR_both_truthy.php.inc:18]: I see. So `if ($value)` is enough, right?

Comment (TomasVotruba) [reply rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_OR_both_truthy.php.inc:18]: It's much less logic to work with than ` if(! is_null( $partPackage ) && ! is_object($partPackage) )`, where is like ~8 operations.

Comment (derrickschoen) [reply rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_OR_both_truthy.php.inc:18]: Is this better?

Comment (TomasVotruba) [reply rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_OR_both_truthy.php.inc:18]: Much better, thanks!

File: rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_OR_both_truthy.php.inc [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        $partPackageList = [ new \StdClass(), 'string' ];
+        $payload = [];
+
+        foreach ($partPackageList as $partPackage) {
+            if (!true) {
+                continue;
+            }
+            if (1 === 1 && 2 === 2) {
+                continue;
+            }
```
Comment (TomasVotruba) [inline rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_OR_both_truthy.php.inc:50]: What is real expected change here? How the correct code should look like?

Comment (derrickschoen) [reply rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_OR_both_truthy.php.inc:50]:
```
            if (!
                (1 === 1 || 2 === 2)
            ) {
                continue;
            }

```

Comment (TomasVotruba) [reply rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_OR_both_truthy.php.inc:50]: Could you update the file with that? So the fixture is correct

Comment (derrickschoen) [reply rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_OR_both_truthy.php.inc]: Done

**Comment by derrickschoen** [comment]
Did I get it?

**Comment by derrickschoen** [comment]
Is this too hacky? I wasn't sure if I should remove the converting OR -> AND altogether.

I'm no CS guy, but I don't think you can replace OR with NOT AND.

Is there a way to go from
```
if(1 === 1 || is_array(1)){
    stuff();
}
```

to

```
if(1 !== 1 && ! is_array(1)){
    continue;
}
stuff();
```

**Comment by TomasVotruba** [comment]
Could you rebase it to 1 commit? It's too many little change back and forth to review.

**Comment by derrickschoen** [comment]
> Could you rebase it to 1 commit? It's too many little change back and forth to review.
> 
> Thank you

Done

File: rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_bool_oprnds.php.inc [hunk 3]
```diff
@@ -45,7 +48,7 @@ class MultiExprsWithBoolOprnds implements RunnableInterface
             if (is_object($partPackage)) {
                 continue;
             }
-            if (1 == '1' && 2 == 't') {
+            if (!(1 == '1' || 2 == 't')) {
```
Comment (TomasVotruba) [inline rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_bool_oprnds.php.inc:51]: Maybe keeping original code from here on would be better. What do you think?

Comment (derrickschoen) [reply rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_bool_oprnds.php.inc:51]: I'm not sure what you mean. If I undo the change, the rector output will not match and the test will not pass.

Comment (TomasVotruba) [reply rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_bool_oprnds.php.inc:51]: Like this:

```diff
foreach( $partPackageList as $partPackage )
        {
-            if(true)
+            if(!true) {
+				continue;
			 }

-             {
                if (1 === 1 || 2 === 2)
                {
                    $payload[] = $partPackage;
                }
-            }
        }
```

**Comment by TomasVotruba** [comment]
I'll resolve this tonight

**Comment by TomasVotruba** [comment]
Merged in master locally: https://github.com/rectorphp/rector/commits/master

**Comment by TomasVotruba** [comment]
Thank you for your work

**Pull request closed without merging**